app/celery/process_letter_client_response_tasks.py

```diff
@@ -0,0 +1,56 @@
+from flask import current_app
+
+from app import notify_celery
+from app.constants import DVLA_NOTIFICATION_DISPATCHED, NOTIFICATION_DELIVERED, NOTIFICATION_TECHNICAL_FAILURE
+from app.dao.notifications_dao import (
+    dao_get_notification_or_history_by_id,
+    dao_update_notification,
+)
+from app.exceptions import NotificationTechnicalFailureException
+
+
+@notify_celery.task(bind=True, name="process-letter-callback")
+def process_letter_callback_data(self, notification_id, page_count, dvla_status):
+    notification = dao_get_notification_or_history_by_id(notification_id)
+
+    check_billable_units_by_id(notification, page_count)
+
+    new_status = determine_new_status(dvla_status)
+
+    if is_duplicate_update(notification.status, new_status):
+        current_app.logger.info(
+            "Duplicate update received for notification id: %s with status: %s",
+            notification_id,
+            new_status,
+        )
+        return
+
+    notification.status = new_status
+
+    dao_update_notification(notification)
+
+    if new_status == NOTIFICATION_TECHNICAL_FAILURE:
+        raise NotificationTechnicalFailureException(
+            f"Letter status received as REJECTED for notification id: {notification.id}"
+        )
+
+
+def check_billable_units_by_id(notification, dvla_page_count):
```

Comment:
method name with suffix __by_id_ suggests that the method relies on _id_ whereas it actually relies on _notification_ entity

Reply:
well spotted, thanks
https://github.com/alphagov/notifications-api/pull/4157/commits/e0eb522c33e3aa06c7a69e951fb4b25643a9116d